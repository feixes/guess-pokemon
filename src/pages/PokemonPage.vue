<template>
  <h1 v-if='!pokemon'>Espere por favor...</h1>
  <div v-else>
    <h1>¿Quien es este pokémon?</h1>
    <PokemonPicture
      :pokemonId="pokemon.id"
      :showPokemon="showPokemon" />
   <PokemonOptions
      :pokemons="pokemonArr" 
      @selection="checkAnswers"/>
    
    <template
      v-if="showAnswer"
      class="fade-in">
      <h2 >{{message}}</h2>
      <button @click="newGame">Nuevo Juego</button>
    </template>
  </div>
</template>

<script>

import PokemonOptions from '@/components/PokemonOptions.vue'
import PokemonPicture from '@/components/PokemonPicture.vue'
import getPokemonOptions from '@/helpers/getPokemonOptions'

export default {
  components:{PokemonOptions, PokemonPicture},
  data() {
    return {
      pokemonArr:[],
      pokemon: null,
      showPokemon: false,
      showAnswer: false,
      message: ''
    }
  },
  methods:{
    async mixPokemonArray(){
      this.pokemonArr = await getPokemonOptions()

      const rndInt = Math.floor(Math.random()*4)

      this.pokemon = this.pokemonArr[rndInt]
    },
    checkAnswers(selectedId){
        //this.showPokemon = this.pokemon.id === pokemonId

        this.showPokemon = true
        this.showAnswer = true

        if (selectedId === this.pokemon.id){
          this.message = `Correcto! Era ${this.pokemon.name}!`
        } else {
          this.message = `Ups! Era ${this.pokemon.name}!`
        }
    },
    newGame(){
      this.showPokemon = false
      this.showAnswer = false
      this.pokemonArr = []
      this.pokemon = null
      this.mixPokemonArray()
    }
  },
  mounted(){
    this.mixPokemonArray()
  }
}
</script>

<style>

</style>