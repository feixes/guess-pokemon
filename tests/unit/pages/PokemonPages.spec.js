
import {shallowMount, mount} from '@vue/test-utils'
import PokemonPage from '@/pages/PokemonPage'
//import PokemonPicture from '@/components/PokemonPicture'
//import PokemonOptions from '@/components/PokemonOptions'
import { pokemons } from '../mocks/pokemons.mock'

describe('PokemonPage', () => { 

    let wrapper

    beforeEach(()=>{
        wrapper = shallowMount(PokemonPage)
    })

    test('debe hacer match con el snapshot', () => { 

        expect(wrapper.html()).toMatchSnapshot()

     })

    test('debe llamar a mixPokemonArray al montar', () => { 
        const mixPokemonArraySpy = jest
                                .spyOn(PokemonPage.methods, 'mixPokemonArray')

        const wrapper = shallowMount(PokemonPage)

        expect(mixPokemonArraySpy).toHaveBeenCalled()
     })

    test('debe de hacer match con el snapshot, una vez cargados los pokemon', ()=>{

        const wrapper = shallowMount(PokemonPage, {
            data(){
                return{
                    pokemonArr: pokemons,
                    pokemon: pokemons[0],
                    showPokemon: false,
                    showAnswer: false,
                    message: ''
                }
            }
        })

        expect(wrapper.html()).toMatchSnapshot()

     })

    test('debe de mostrar los componentes PokemonPicture y PokemonOptions', () => { 

        const wrapper = shallowMount(PokemonPage, {
            data(){
                return{
                    pokemonArr: pokemons,
                    pokemon: pokemons[0],
                    showPokemon: false,
                    showAnswer: false,
                    message: ''
                }
            }
        })

        const picture = wrapper.findComponent({name: 'PokemonPicture'})
        const options = wrapper.findComponent({name: 'PokemonOptions'})
        
        // Otra forma, buscar el stub del componente que se crea del shallowMount
        //const picture = wrapper.find('pokemon-picture-stub')

        expect(picture.exists()).toBe(true)
        expect(options.exists()).toBe(true)

        expect(picture.vm.pokemonId).toBe(1)

        expect(options.vm.pokemons).toBeTruthy()

     })

    test('pruebas con checkAnser', async ()=>{
        const wrapper = shallowMount(PokemonPage, {
            data(){
                return{
                    pokemonArr: pokemons,
                    pokemon: pokemons[0],
                    showPokemon: false,
                    showAnswer: false,
                    message: ''
                }
            }
        })

        await wrapper.vm.checkAnswers(1)

        expect(wrapper.find('h2').exists()).toBe(true)
        expect(wrapper.vm.showPokemon).toBe(true)
        expect(wrapper.find('h2').text()).toBe(`Correcto! Era ${pokemons[0].name}!`)
        expect(wrapper.vm.message).toBe(`Ups! Era ${pokemons[0].name}!`)

     })

 })