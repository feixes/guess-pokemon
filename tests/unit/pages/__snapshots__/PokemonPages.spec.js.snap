// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PokemonPage debe de hacer match con el snapshot, una vez cargados los pokemon 1`] = `
<div>
  <h1>¿Quien es este pokémon?</h1>
  <pokemon-picture-stub pokemonid="1" showpokemon="false"></pokemon-picture-stub>
  <pokemon-options-stub pokemons="[object Object],[object Object],[object Object],[object Object]"></pokemon-options-stub>
  <!--v-if-->
</div>
`;

exports[`PokemonPage debe hacer match con el snapshot 1`] = `<h1>Espere por favor...</h1>`;
